File: static/code/agent-templates/python/recommendation_agent.py
'''
Redis Recommendation Engine

To run this code:
    Install dependencies:
        pip install pandas redisvl[all] redis openai

    Set environment variables:
        export LLM_API_KEY=your_${formData.llmModel.toLowerCase()}_api_key
        export LLM_API_BASE_URL=your_${formData.llmModel.toLowerCase()}_api_base_url
            (optional - default: ${CONFIG.models[formData.llmModel].baseUrl})
        export LLM_MODEL=your_${formData.llmModel.toLowerCase()}_model 
            (optional - default: ${CONFIG.models[formData.llmModel].defaultModel})
        export REDIS_HOST=your_redis_host
        export REDIS_PORT=your_redis_port
        export REDIS_PASSWORD=your_redis_password

The datasets are downloadable from here:

- https://redis-ai-resources.s3.us-east-2.amazonaws.com/recommenders/datasets/collaborative-filtering/ratings_small.csv and
- https://redis-ai-resources.s3.us-east-2.amazonaws.com/recommenders/datasets/collaborative-filtering/movies_metadata.csv

Download those files and place them in the 'datasets/collaborative_filtering/' directory, local to your code.
'''
import os

import openai
import pandas as pd
import redis

from redisvl.schema import IndexSchema
from redisvl.index import SearchIndex
from redisvl.query import FilterQuery
from redisvl.query.filter import Num

class RecommendationAgent:
    def __init__(self, session_name="movie_recommendations"):
        # Validate environment variables
        self.llm_api_key = os.getenv('LLM_API_KEY')
        if not self.llm_api_key:
            raise ValueError("LLM_API_KEY environment variable is required")

        self.llm_base_url = os.getenv('LLM_API_BASE_URL', '${CONFIG.models[formData.llmModel].baseUrl}')
        self.llm_model = os.getenv('LLM_MODEL', '${CONFIG.models[formData.llmModel].defaultModel}')

        # Connect to Redis with error handling
        try:
            self.redis_client = redis.Redis(
                host=os.getenv('REDIS_HOST', 'localhost'),
                port=int(os.getenv('REDIS_PORT', 6379)),
                username=os.getenv('REDIS_USERNAME', 'default'),
                password=os.getenv('REDIS_PASSWORD', ''),
                decode_responses=True,
                socket_connect_timeout=5,
                socket_timeout=5
            )
            # Test Redis connection
            self.redis_client.ping()
            print("Connected to Redis successfully")

        except redis.ConnectionError as e:
            print(f"Failed to connect to Redis: {e}")
            print("Please check your Redis connection settings and ensure Redis is running.")
            raise
        except Exception as e:
            print(f"Redis connection error: {e}")
            raise

        # Initialize LLM client with error handling
        try:
            self.client = openai.OpenAI(api_key=self.llm_api_key, base_url=self.llm_base_url)
            # Test LLM connection with a simple call
            test_response = self.client.chat.completions.create(
                model=self.llm_model,
                messages=[{"role": "user", "content": "Hello"}],
                max_tokens=5
            )
            print("Connected to LLM successfully")

        except openai.AuthenticationError:
            print("LLM authentication failed. Please check your API key.")
            raise
        except Exception as e:
            print(f"LLM connection error: {e}")
            raise

        # Initialize the movie index
        self.index = None
        self.movies_df = None
        self._setup_movie_index()

    def _setup_movie_index(self):
        """Load movie data and create Redis index"""
        try:
            print("Loading movie datasets...")

            # Check if data files exist
            ratings_file = 'datasets/collaborative_filtering/ratings_small.csv'
            movies_file = 'datasets/collaborative_filtering/movies_metadata.csv'

            if not os.path.exists(ratings_file):
                raise FileNotFoundError(f"Ratings dataset not found: {ratings_file}")
            if not os.path.exists(movies_file):
                raise FileNotFoundError(f"Movies dataset not found: {movies_file}")

            # Load and prepare data
            ratings_df = pd.read_csv(ratings_file)[['userId', 'movieId', 'rating']]
            movies_df = pd.read_csv(movies_file)[['id', 'title', 'genres', 'revenue']].dropna()

            if ratings_df.empty or movies_df.empty:
                raise ValueError("One or more datasets are empty")

            print(f"Loaded {len(ratings_df)} ratings and {len(movies_df)} movies")

            # Calculate movie popularity metrics
            movie_stats = ratings_df.groupby('movieId').agg({
                'rating': ['count', 'mean']
            }).round(2)
            movie_stats.columns = ['rating_count', 'avg_rating']
            movie_stats['popularity_score'] = movie_stats['rating_count'] * movie_stats['avg_rating']
            movie_stats = movie_stats.reset_index()

            # Merge with movie metadata
            movies_df['movieId'] = movies_df['id'].astype(str)
            movie_stats['movieId'] = movie_stats['movieId'].astype(str)
            self.movies_df = movies_df.merge(movie_stats, on='movieId', how='inner')

            if self.movies_df.empty:
                raise ValueError("No movies found after merging ratings and metadata")

            print(f"Processed {len(self.movies_df)} movies with ratings")

            # Create RedisVL index for popularity-based search
            schema = IndexSchema.from_dict({
                'index': {'name': 'movies', 'prefix': 'movie', 'storage_type': 'json'},
                'fields': [
                    {'name': 'movieId', 'type': 'tag'},
                    {'name': 'title', 'type': 'text'},
                    {'name': 'genres', 'type': 'tag'},
                    {'name': 'revenue', 'type': 'numeric'},
                    {'name': 'rating_count', 'type': 'numeric'},
                    {'name': 'avg_rating', 'type': 'numeric'},
                    {'name': 'popularity_score', 'type': 'numeric'}
                ]
            })

            print("Creating Redis search index...")
            self.index = SearchIndex(schema, redis_client=self.redis_client)
            self.index.create(overwrite=True)
            self.index.load(self.movies_df.to_dict(orient='records'))

            print("Movie recommendation system initialized successfully!")

        except FileNotFoundError as e:
            print(f"Data files not found: {e}")
            print("Please ensure the movie datasets are available in the 'datasets/collaborative_filtering/' directory")
            print("Expected files:")
            print("  - datasets/collaborative_filtering/ratings_small.csv")
            print("  - datasets/collaborative_filtering/movies_metadata.csv")
            self.index = None
        except pd.errors.EmptyDataError as e:
            print(f"Data file is empty or corrupted: {e}")
            self.index = None
        except Exception as e:
            print(f"Error setting up movie index: {e}")
            self.index = None

    def get_popular_movies(self, filters=None, num_results=10, sort_by='popularity_score', sort_order='desc'):
        """Get popular movies based on filters and sorting criteria

        Args:
            filters: Filter expression for the query
            num_results: Number of results to return
            sort_by: Field to sort by ('popularity_score', 'avg_rating', 'rating_count', 'revenue')
            sort_order: Sort order ('asc' for ascending, 'desc' for descending)
        """
        if not self.index:
            return []

        # Create the query
        query = FilterQuery(
            filter_expression=filters,
            return_fields=['title', 'genres', 'rating_count', 'avg_rating', 'popularity_score'],
            num_results=num_results
        )

        # Add sorting with specified order
        asc_order = sort_order.lower() == 'asc'
        query.sort_by(sort_by, asc=asc_order)

        results = self.index.query(query)
        return [(r['title'], r['genres'], r['rating_count'], r['avg_rating'], r['popularity_score'])
                for r in results]

    def get_top_rated_movies(self, min_reviews=50, num_results=10, sort_order='desc'):
        """Get highest rated movies with minimum review count"""
        filters = Num('rating_count') >= min_reviews
        return self.get_popular_movies(filters=filters, num_results=num_results, sort_by='avg_rating', sort_order=sort_order)

    def get_blockbuster_movies(self, min_revenue=50_000_000, num_results=10, sort_order='desc'):
        """Get blockbuster movies with high revenue"""
        filters = Num('revenue') >= min_revenue
        return self.get_popular_movies(filters=filters, num_results=num_results, sort_by='revenue', sort_order=sort_order)

    def get_hidden_gems(self, max_reviews=200, min_rating=7.5, num_results=10, sort_order='desc'):
        """Get lesser-known but highly rated movies"""
        filters = (Num('rating_count') <= max_reviews) & (Num('avg_rating') >= min_rating)
        return self.get_popular_movies(filters=filters, num_results=num_results, sort_by='avg_rating', sort_order=sort_order)

    def get_movies_by_genre(self, genre, num_results=10, sort_by='popularity_score', sort_order='desc'):
        """Get movies filtered by genre (post-processing)"""
        all_movies = self.get_popular_movies(num_results=num_results*3, sort_by=sort_by, sort_order=sort_order)
        genre_movies = []

        for movie in all_movies:
            movie_genres = movie[1].lower() if movie[1] else ""
            if genre.lower() in movie_genres:
                genre_movies.append(movie)
                if len(genre_movies) >= num_results:
                    break

        return genre_movies

    def cleanup(self):
        """Clean up Redis index"""
        if self.index:
            self.index.clear()

    def _parse_user_query(self, user_query: str) -> dict:
        """Use LLM to parse user query and extract recommendation parameters"""
        try:
            system_prompt = """You are a movie recommendation assistant. Parse the user's query and extract relevant parameters for movie filtering.

Return a JSON object with these fields (use null if not specified):
- "genres": array of genre strings (e.g., ["Action", "Comedy"])
- "min_rating": minimum average rating (0-10)
- "min_reviews": minimum number of reviews
- "max_results": number of recommendations (default 5, max 10)
- "sort_by": "popularity_score", "avg_rating", "rating_count", or "revenue"
- "sort_order": "desc" for descending (highest first), "asc" for ascending (lowest first). Default "desc"
- "revenue_filter": "blockbuster" for high-revenue movies, null otherwise

Examples:
- "action movies" -> {"genres": ["Action"], "min_rating": null, "min_reviews": null, "max_results": 5, "sort_by": "popularity_score", "sort_order": "desc", "revenue_filter": null}
- "highly rated comedies with at least 100 reviews" -> {"genres": ["Comedy"], "min_rating": 7.0, "min_reviews": 100, "max_results": 5, "sort_by": "avg_rating", "sort_order": "desc", "revenue_filter": null}
- "lowest rated movies" -> {"genres": null, "min_rating": null, "min_reviews": null, "max_results": 5, "sort_by": "avg_rating", "sort_order": "asc", "revenue_filter": null}
- "popular blockbuster movies" -> {"genres": null, "min_rating": null, "min_reviews": 50, "max_results": 5, "sort_by": "popularity_score", "sort_order": "desc", "revenue_filter": "blockbuster"}"""

            response = self.client.chat.completions.create(
                model=self.llm_model,
                messages=[
                    {"role": "system", "content": system_prompt},
                    {"role": "user", "content": user_query}
                ],
                temperature=0.1
            )

            import json
            params = json.loads(response.choices[0].message.content)
            return params

        except Exception as e:
            print(f"Error parsing query: {e}")
            # Return default parameters
            return {"genres": None, "min_rating": None, "min_reviews": None, "max_results": 5, "sort_by": "popularity_score", "sort_order": "desc", "revenue_filter": None}

    def recommend_movies(self, user_query: str) -> str:
        """Process user query and return movie recommendations"""
        if not self.index:
            return "Sorry, the movie database is not available. Please check that the data files are present."

        try:
            # Parse user query
            params = self._parse_user_query(user_query)

            # Build filters based on parsed parameters
            filters = []

            if params.get("min_rating"):
                filters.append(Num('avg_rating') >= params["min_rating"])

            if params.get("min_reviews"):
                filters.append(Num('rating_count') >= params["min_reviews"])

            if params.get("revenue_filter") == "blockbuster":
                filters.append(Num('revenue') > 30_000_000)

            # Combine filters
            combined_filter = None
            if filters:
                combined_filter = filters[0]
                for f in filters[1:]:
                    combined_filter = combined_filter & f

            # Get recommendations
            num_results = min(params.get("max_results", 5), 10)
            sort_by = params.get("sort_by", "popularity_score")
            sort_order = params.get("sort_order", "desc")

            movies = self.get_popular_movies(
                filters=combined_filter,
                num_results=num_results,
                sort_by=sort_by,
                sort_order=sort_order
            )

            # Filter by genre if specified (post-processing since RedisVL genre filtering can be complex)
            if params.get("genres"):
                genre_keywords = [g.lower() for g in params["genres"]]
                filtered_movies = []
                for movie in movies:
                    movie_genres = movie[1].lower() if movie[1] else ""
                    if any(keyword in movie_genres for keyword in genre_keywords):
                        filtered_movies.append(movie)
                movies = filtered_movies[:num_results]

            if not movies:
                return "Sorry, no movies found matching your criteria. Try adjusting your preferences."

            # Format response
            response = f"Based on your request '{user_query}', here are my recommendations:\n\n"
            for i, (title, genres, rating_count, avg_rating, popularity_score) in enumerate(movies, 1):
                response += f"{i}. {title}\n"
                response += f"   Genres: {genres}\n"
                response += f"   Average Rating: {float(avg_rating):.1f}/10 ({int(rating_count)} reviews)\n"
                response += f"   Popularity Score: {float(popularity_score):.1f}\n\n"

            return response

        except Exception as e:
            return f"Sorry, there was an error processing your request: {e}"

if __name__ == "__main__":
    try:
        print("Initializing Redis Movie Recommendation Agent...")
        agent = RecommendationAgent()

        if not agent.index:
            print("Failed to initialize the recommendation system.")
            print("Please check your data files and Redis connection, then try again.")
            exit(1)

        print("\nWelcome to the Redis Movie Recommendation Agent!")
        print("Ask me for movie recommendations and I'll help you find something great to watch.")
        print("Type 'quit' or 'exit' to stop.\n")

        # Demo some initial recommendations
        print("Here's a quick demo of what I can do:")
        print(agent.recommend_movies("Show me some popular movies"))

        # Interactive loop
        while True:
            try:
                user_input = input("\nWhat kind of movies are you looking for? ")

                if user_input.lower() in ['quit', 'exit', 'bye']:
                    print("Thanks for using the movie recommendation agent! Goodbye!")
                    break

                if user_input.strip():
                    response = agent.recommend_movies(user_input)
                    print(f"\n{response}")
                else:
                    print("Please enter a movie preference or type 'quit' to exit.")

            except KeyboardInterrupt:
                print("\n\nGoodbye!")
                break
            except Exception as e:
                print(f"An error occurred: {e}")
                print("Please try again or type 'quit' to exit.")

        # Cleanup
        agent.cleanup()

    except ValueError as e:
        print(f"Configuration error: {e}")
        print("Please check your environment variables and try again.")
        exit(1)
    except Exception as e:
        print(f"Failed to initialize the recommendation agent: {e}")
        exit(1)

File: static/code/agent-templates/javascript/recommendation_agent.js
/*
 * Redis Recommendation Engine (Node.js)
 * Uses node-redis with Redis Search for movie recommendations
 *
 * To run this code:
 *   Install dependencies:
 *     npm install redis openai dotenv csv-parse
 *
 *   Set environment variables:
 *     LLM_API_KEY=your_${formData.llmModel.toLowerCase()}_api_key
 *     LLM_API_BASE_URL=your_base_url  (optional, default: ${CONFIG.models[formData.llmModel].baseUrl})
 *     LLM_MODEL=your_model_name       (optional, default: ${CONFIG.models[formData.llmModel].defaultModel})
 *     REDIS_URL=redis://localhost:6379
 *       (or use REDIS_HOST, REDIS_PORT, REDIS_PASSWORD, REDIS_USERNAME separately)
 *
 * Download datasets (all three are required):
 *   https://redis-ai-resources.s3.us-east-2.amazonaws.com/recommenders/datasets/collaborative-filtering/ratings_small.csv
 *   https://redis-ai-resources.s3.us-east-2.amazonaws.com/recommenders/datasets/collaborative-filtering/movies_metadata.csv
 *   https://redis-ai-resources.s3.us-east-2.amazonaws.com/recommenders/datasets/collaborative-filtering/links.csv
 * Place them in datasets/collaborative_filtering/ relative to this file.
 * links.csv bridges MovieLens movieIds (ratings) to TMDB ids (movies_metadata).
 */

require('dotenv').config();
const { createClient } = require('redis');
const OpenAI = require('openai');
const { parse } = require('csv-parse/sync');
const fs = require('fs');
const path = require('path');

const INDEX_NAME   = 'movies_idx';
const MOVIE_PREFIX  = 'movie:';
const LOAD_SENTINEL = 'movies:load_complete';

const CONFIG = {
  maxResults:       10,
  defaultResults:   5,
  minRevenueFilter: 30_000_000,
  validSortFields:  new Set(['popularityScore', 'avgRating', 'ratingCount', 'revenue']),
  validSortOrders:  new Set(['DESC', 'ASC']),
};

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

/**
 * Parse genres from movies_metadata.csv.
 * The field is stored as a Python-style list of dicts, e.g.:
 *   "[{'id': 28, 'name': 'Action'}, {'id': 12, 'name': 'Adventure'}]"
 * Returns a comma-separated string for use as a Redis TAG field.
 */
function parseGenres(raw) {
  if (!raw || raw === '[]') return '';
  try {
    const json = raw.replace(/'/g, '"').replace(/None/g, 'null').replace(/True/g, 'true').replace(/False/g, 'false');
    const parsed = JSON.parse(json);
    return parsed.map((g) => g?.name).filter(Boolean).join(',');
  } catch {
    console.warn('parseGenres: could not parse genres field, storing empty string. Raw value:', raw?.slice(0, 80));
    return '';
  }
}

function safeNumber(value, fallback = 0) {
  const n = Number(value);
  return isFinite(n) ? n : fallback;
}

/**
 * Validate and sanitize LLM-returned query params.
 * Rejects any field that doesn't match expected types or allowed values.
 */
function validateQueryParams(raw) {
  return {
    genres: Array.isArray(raw?.genres)
      ? raw.genres.filter((g) => typeof g === 'string' && g.trim())
      : null,
    minRating: typeof raw?.minRating === 'number' && raw.minRating >= 0 && raw.minRating <= 5
      ? raw.minRating
      : null,
    minReviews: typeof raw?.minReviews === 'number' && raw.minReviews > 0
      ? Math.floor(raw.minReviews)
      : null,
    maxResults: typeof raw?.maxResults === 'number'
      ? Math.min(Math.max(1, Math.floor(raw.maxResults)), CONFIG.maxResults)
      : CONFIG.defaultResults,
    sortBy: CONFIG.validSortFields.has(raw?.sortBy)
      ? raw.sortBy
      : 'popularityScore',
    sortOrder: CONFIG.validSortOrders.has(raw?.sortOrder?.toUpperCase?.())
      ? raw.sortOrder.toUpperCase()
      : 'DESC',
    revenueFilter: raw?.revenueFilter === true,
  };
}

// ---------------------------------------------------------------------------
// Agent class
// ---------------------------------------------------------------------------

class RecommendationAgent {
  constructor() {
    // For local providers (e.g. Ollama), any non-empty string works. For hosted providers, use your real key.
    this.llmApiKey = process.env.LLM_API_KEY || 'no-key-needed';

    this.llmBaseUrl = process.env.LLM_API_BASE_URL || '${CONFIG.models[formData.llmModel].baseUrl}';
    this.llmModel   = process.env.LLM_MODEL         || '${CONFIG.models[formData.llmModel].defaultModel}';

    this.openai = new OpenAI({ apiKey: this.llmApiKey, baseURL: this.llmBaseUrl });
    this.redisClient = null;
    this.indexReady = false;
  }

  async connect() {
    const clientOptions = process.env.REDIS_URL
      ? { url: process.env.REDIS_URL }
      : {
          socket: {
            host: process.env.REDIS_HOST || 'localhost',
            port: parseInt(process.env.REDIS_PORT) || 6379,
          },
          password: process.env.REDIS_PASSWORD || undefined,
          username: process.env.REDIS_USERNAME  || 'default',
        };

    this.redisClient = createClient(clientOptions);
    this.redisClient.on('error', (err) => console.error('Redis error:', err));
    await this.redisClient.connect();
    console.log('Connected to Redis successfully');
    console.log('LLM configured:', this.llmModel);

    await this._setupMovieIndex();
  }

  async _indexExists() {
    try {
      // Check the sentinel written only after a full successful load.
      // num_docs > 0 alone is not sufficient — a partial import looks non-empty
      // but leaves the index permanently incomplete on subsequent startups.
      return (await this.redisClient.exists(LOAD_SENTINEL)) === 1;
    } catch {
      return false;
    }
  }

  async _setupMovieIndex() {
    // Skip loading if the index already exists and has documents.
    if (await this._indexExists()) {
      console.log('Movie index already loaded, skipping dataset import.');
      this.indexReady = true;
      return;
    }

    const ratingsFile = path.join('datasets', 'collaborative_filtering', 'ratings_small.csv');
    const moviesFile  = path.join('datasets', 'collaborative_filtering', 'movies_metadata.csv');
    const linksFile   = path.join('datasets', 'collaborative_filtering', 'links.csv');

    for (const f of [ratingsFile, moviesFile, linksFile]) {
      if (!fs.existsSync(f)) {
        console.warn(`Movie dataset not found: ${f}. Skipping index setup.`);
        return;
      }
    }

    console.log('Loading movie datasets...');

    let ratings, movies, links;
    try {
      ratings = parse(fs.readFileSync(ratingsFile), { columns: true, cast: true });
      movies  = parse(fs.readFileSync(moviesFile),  { columns: true, cast: true });
      links   = parse(fs.readFileSync(linksFile),   { columns: true, cast: true });
    } catch (err) {
      console.error('Failed to parse dataset files:', err.message);
      return;
    }

    if (!ratings.length || !movies.length || !links.length) {
      console.error('One or more dataset files are empty.');
      return;
    }

    // Aggregate ratings per MovieLens movieId
    const stats = {};
    for (const r of ratings) {
      if (!r.movieId || !isFinite(r.rating)) continue;
      if (!stats[r.movieId]) stats[r.movieId] = { count: 0, total: 0 };
      stats[r.movieId].count++;
      stats[r.movieId].total += r.rating;
    }

    // Build MovieLens movieId → TMDB id bridge via links.csv.
    // ratings_small uses MovieLens ids; movies_metadata uses TMDB ids — they are
    // different id spaces and cannot be joined directly without this mapping.
    const movieLensToTmdb = new Map();
    for (const row of links) {
      if (row.movieId && row.tmdbId) {
        movieLensToTmdb.set(String(row.movieId), String(row.tmdbId));
      }
    }

    // Index movies_metadata by TMDB id for O(1) lookup
    const moviesByTmdb = new Map();
    for (const m of movies) {
      if (m.id) moviesByTmdb.set(String(m.id), m);
    }

    // Join ratings → links → metadata
    const merged = [];
    for (const [movieLensId, s] of Object.entries(stats)) {
      const tmdbId = movieLensToTmdb.get(movieLensId);
      if (!tmdbId) continue;
      const m = moviesByTmdb.get(tmdbId);
      if (!m || !m.title) continue;
      const avgRating = s.total / s.count;
      merged.push({
        movieId:         movieLensId,
        title:           String(m.title).trim(),
        genres:          parseGenres(m.genres),
        revenue:         safeNumber(m.revenue),
        ratingCount:     s.count,
        avgRating:       Math.round(avgRating * 100) / 100,
        popularityScore: Math.round(s.count * avgRating * 100) / 100,
      });
    }

    if (!merged.length) {
      console.error('No valid movies found after merging datasets.');
      return;
    }

    console.log(`Processed ${merged.length} movies`);

    // Drop existing index and its documents so stale movie keys don't survive the reload.
    const indexExists = await this.redisClient.ft.info(INDEX_NAME).then(() => true).catch(() => false);
    if (indexExists) await this.redisClient.ft.dropIndex(INDEX_NAME, { DD: true });

    await this.redisClient.ft.create(
      INDEX_NAME,
      {
        '$.movieId':         { type: 'TAG',     AS: 'movieId' },
        '$.title':           { type: 'TEXT',    AS: 'title' },
        '$.genres':          { type: 'TAG',     AS: 'genres', SEPARATOR: ',' },
        '$.revenue':         { type: 'NUMERIC', AS: 'revenue' },
        '$.ratingCount':     { type: 'NUMERIC', AS: 'ratingCount' },
        '$.avgRating':       { type: 'NUMERIC', AS: 'avgRating' },
        '$.popularityScore': { type: 'NUMERIC', AS: 'popularityScore' },
      },
      { ON: 'JSON', PREFIX: MOVIE_PREFIX }
    );

    // Load using a pipeline for efficiency
    const pipeline = this.redisClient.multi();
    for (const movie of merged) {
      pipeline.json.set(`${MOVIE_PREFIX}${movie.movieId}`, '$', movie);
    }
    await pipeline.exec();

    // Write the sentinel only after all documents are loaded. _indexExists checks
    // this key so a partial import (pipeline fails midway) never looks complete.
    await this.redisClient.set(LOAD_SENTINEL, '1');

    this.indexReady = true;
    console.log('Movie recommendation system initialized successfully!');
  }

  async _parseUserQuery(userQuery) {
    const systemPrompt = `You are a movie recommendation assistant. Parse the user's query and return a JSON object with:
- "genres": array of genre name strings or null
- "minRating": minimum average rating (0-5 star scale) or null
- "minReviews": minimum review count or null
- "maxResults": number of results (default 5, max 10)
- "sortBy": one of "popularityScore", "avgRating", "ratingCount", "revenue"
- "sortOrder": "DESC" or "ASC"
- "revenueFilter": true for blockbusters, null otherwise

Return only valid JSON with no explanation or markdown.`;

    try {
      const response = await this.openai.chat.completions.create({
        model: this.llmModel,
        messages: [
          { role: 'system', content: systemPrompt },
          { role: 'user', content: userQuery },
        ],
        temperature: 0.1,
      });

      const raw = JSON.parse(response.choices[0]?.message?.content || '{}');
      return validateQueryParams(raw); // always returns a safe, validated object
    } catch {
      return validateQueryParams({}); // safe defaults on any failure
    }
  }

  async recommendMovies(userQuery) {
    if (!this.indexReady) {
      return 'The movie database is not available. Please check that the dataset files are present.';
    }

    const params = await this._parseUserQuery(userQuery);

    // Build filter parts — genres now filtered in Redis via TAG, not in JavaScript
    const filterParts = [];
    if (params.minRating != null)  filterParts.push(`@avgRating:[${params.minRating} +inf]`);
    if (params.minReviews != null) filterParts.push(`@ratingCount:[${params.minReviews} +inf]`);
    if (params.revenueFilter) filterParts.push(`@revenue:[${CONFIG.minRevenueFilter} +inf]`);
    if (params.genres?.length) {
      // Redis TAG filter: match any of the requested genres.
      // Hyphens (e.g. "Film-Noir") and spaces (e.g. "Science Fiction") are stored intact
      // at ingest, so they must be preserved and backslash-escaped in the query rather
      // than stripped, otherwise hyphenated and multi-word genres never match.
      const tagList = params.genres
        .map((g) => g.replace(/[^a-zA-Z0-9 \-]/g, '').trim()
          .replace(/-/g, '\\-')
          .replace(/ +/g, '\\ '))
        .filter(Boolean)
        .join('|');
      if (tagList) filterParts.push(`@genres:{${tagList}}`);
    }

    const filterQuery = filterParts.length > 0 ? filterParts.join(' ') : '*';

    let results;
    try {
      results = await this.redisClient.ft.search(INDEX_NAME, filterQuery, {
        RETURN: ['title', 'genres', 'ratingCount', 'avgRating', 'popularityScore', 'revenue'],
        SORTBY: { BY: params.sortBy, DIRECTION: params.sortOrder },
        LIMIT: { from: 0, size: params.maxResults },
      });
    } catch (err) {
      console.error('Search error:', err.message);
      return 'Sorry, there was an error searching the movie database.';
    }

    const movies = results?.documents?.map((d) => d.value).filter(Boolean) ?? [];

    if (!movies.length) {
      return "Sorry, no movies found matching your criteria. Try adjusting your preferences.";
    }

    let response = `Based on your request '${userQuery}', here are my recommendations:\n\n`;
    movies.forEach((m, i) => {
      response += `${i + 1}. ${m.title}\n`;
      response += `   Genres: ${m.genres || 'N/A'}\n`;
      response += `   Average Rating: ${parseFloat(m.avgRating || 0).toFixed(1)}/5 (${m.ratingCount || 0} reviews)\n`;
      response += `   Popularity Score: ${parseFloat(m.popularityScore || 0).toFixed(1)}\n`;
      if (m.revenue && parseFloat(m.revenue) > 0) {
        response += `   Box Office: $${(parseFloat(m.revenue) / 1_000_000).toFixed(0)}M\n`;
      }
      response += '\n';
    });
    return response;
  }

  async disconnect() {
    if (this.redisClient) await this.redisClient.disconnect();
  }
}

async function main() {
  const agent = new RecommendationAgent();
  try {
    await agent.connect();
  } catch (err) {
    console.error('Failed to initialize agent:', err.message);
    await agent.disconnect();
    process.exit(1);
  }

  console.log('\nWelcome to the Redis Movie Recommendation Agent!');
  console.log("Ask for movie recommendations. Type 'quit' to exit.\n");
  console.log("Here's a quick demo:");
  console.log(await agent.recommendMovies('Show me some popular movies'));

  const readline = require('readline');
  const rl = readline.createInterface({ input: process.stdin, output: process.stdout });

  const askQuestion = () => {
    rl.question('\nWhat kind of movies are you looking for? ', async (input) => {
      if (['quit', 'exit', 'bye'].includes(input.toLowerCase())) {
        console.log('Goodbye!');
        rl.close();
        await agent.disconnect();
        return;
      }
      if (input.trim()) {
        try {
          console.log('\n' + await agent.recommendMovies(input));
        } catch (err) {
          console.error('Error:', err.message);
        }
      }
      askQuestion();
    });
  };
  askQuestion();
}

main();
